### ISSUE TITLE
Question: Reproducing VR Controller Tracking Task

### ISSUE THREAD

**Issue by frredy99** [ZhengyiLuo/PULSE#17]
@ZhengyiLuo Hello!
Thanks for sharing your work!

I have been trying to reproduce the VR tracking task result, and there are some questions related to it.

### 1. Motion Dataset for Training
In the paper, it seems VR tracking task uses the same amass train set as of when training motion imitation.
![image](https://github.com/user-attachments/assets/625f9baf-2096-4e3a-bdc2-2c9f0513f0d4)
However, the VR Controller Tracking command in the README does not explicitly mention which motion data file is used. Is it correct to use the amass train set (`amass_test_take6_upright.pkl` which is generated from `python scripts/data_process/convert_amass_data.py` and used for PULSE motion imitation training)?

### 2. Configs from `pulse_z_task.yaml`
The README guides to use `pulse_z_task.yaml` file for VR tracking task. However, it seems some configs does not match with the paper; first of all, `val: -1` for `sigma_init`. In the paper, task policy uses fixed variance of 0.22.
![image](https://github.com/user-attachments/assets/7ce15fd3-12e6-4b97-b7ed-f81b79db8c5e)
However, exp(-1)=0.3679 does not match. Isn't `val: -1.5` correct, since exp(-1.5)=0.2231?
Also, `units: [2048, 1024, 512]` for `mlp` doesn't match with the paper;
![image](https://github.com/user-attachments/assets/3b07883f-4c34-4d99-9e08-9c4f20d3247c)

### 3. Number of Samples for VR Task
In the above picture, number of samples for converging when training speed, reach, strike and terrain task is mentioned. However, I could not find VR task case. Could you please guide how many samples needed for VR task to converge?

Thanks in advance.

**Comment by ZhengyiLuo** [comment]
Hi! 

1. For training, the VR tracking task uses the same training dataset as motion imitation. It should be `amass_train_take6_upright.pkl`. 

2. `pulse_z_task` is for training the downstream tasks, which uses `units: [2048, 1024, 512]`. For VR controller tracking task, I use a slightly different config with `units: [2048, 1536, 1024, 1024, 512, 512]` and `val: -1.5`. I just pushed the modified yaml file. Thanks for pointing it out! 

3. VR controller task I trained for around 2 x 10^9 (22500 epochs) samples. Training for longer does yield marginally better results though.

**Comment by frredy99** [comment]
Thanks for answering!!

Though I have tried with MLP size of `units: [2048, 1536, 1024, 1024, 512, 512]` and `val: -1.5`, but still, the evaluation metrics couldn't reach that of the paper. For instance, evaluated on the AMASS test set, success rate: 73.9%, mpjpe: 79.2 for the model I trained myself, whereas the paper reported success rate: 93.4%, mpjpe: 67.7.

Trying to catch out my possible mistakes, I found out this sentence in the appendix C.4 section in the paper.
"We use the same progressive training procedure for training the tracking policy."

![image](https://github.com/user-attachments/assets/ea966f0e-45c1-46ee-835f-6846a1b3c219)

Does "progressive training procedure" stand for the PNN architecture which was used in PHC?
If so, I think that would be the reason of failing to reproduce the evaluation metrics.

\+ For VR task, `num_envs` is 1536 and `horizon_length` is 32, so there would be 1536x32 samples per epoch. Then, doesn't required epoch should be around 40,000 ( = 2x10^9 / (1536x32) )??


Thank you and best regards,

**Comment by ZhengyiLuo** [comment]
> Thanks for answering!!
> 
> Though I have tried with MLP size of `units: [2048, 1536, 1024, 1024, 512, 512]` and `val: -1.5`, but still, the evaluation metrics couldn't reach that of the paper. For instance, evaluated on the AMASS test set, success rate: 73.9%, mpjpe: 79.2 for the model I trained myself, whereas the paper reported success rate: 93.4%, mpjpe: 67.7.
> 
> Trying to catch out my possible mistakes, I found out this sentence in the appendix C.4 section in the paper. "We use the same progressive training procedure for training the tracking policy."
> 
> ![image](https://private-user-images.githubusercontent.com/93816728/398690427-ea966f0e-45c1-46ee-835f-6846a1b3c219.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYyMzU3MjcsIm5iZiI6MTczNjIzNTQyNywicGF0aCI6Ii85MzgxNjcyOC8zOTg2OTA0MjctZWE5NjZmMGUtNDVjMS00NmVlLTgzNWYtNjg0NmExYjNjMjE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA3VDA3MzcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkYzFhZmY3Y2M2ZTRlOGM4OTgzMDU2YjQ2NWEwYjhkZjI5MTMzMzQzMTNiYjRkMTg5OTM3NTM0YzE3YjNmOTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IzT6wq1hVZ6cUrC9W4sHRpzSn2FZZ6yCSLncT6mvK2c)
> 
> Does "progressive training procedure" stand for the PNN architecture which was used in PHC? If so, I think that would be the reason of failing to reproduce the evaluation metrics.
> 
> + For VR task, `num_envs` is 1536 and `horizon_length` is 32, so there would be 1536x32 samples per epoch. Then, doesn't required epoch should be around 40,000 ( = 2x10^9 / (1536x32) )??
> 
> Thank you and best regards,

I double checked and realized that for VR controller tracking task I trained using 3072 envs; so 25500 epoches is roughly correct. 

The progressive training procedure is referring to the one used in this paper (PULSE) instead of PNN. It just refers to the automatic hard-negative mining procedure. **I realized this was not turned on (the auto_pmcp_soft) flag** in the config file I released; my bad, I have pushed the modified code. 

How long is your model trained for? What is the training set success rate for your tracking model right now? For AMASS the training set distribution covers the test set pretty well so training set success rate is usually a pretty good indicator.

**Comment by frredy99** [comment]
Thank you for answering!

#### 1. How long is your model trained for?
I trained with previous config (1536 envs), so total 40,000 epochs to match 2 x 10^9 samples.

#### 2. What is the training set success rate for your tracking model right now?
||success rate|g-mpjpe|mpjpe|acc|vel|
|-|------|---|---|--|--|
|Train set|0.96213|59.30511|51.59011|4.02125|6.09905|
|Validation set|0.8|84.7218|71.78029|6.72136|10.5008|
|Evaluation set|0.67391|116.4351|81.12007|8.69799|13.62242|

\* All 3 sets are created with [convert_amass_data.py](https://github.com/ZhengyiLuo/PULSE/blob/master/scripts/data_process/convert_amass_data.py), changing [`process_split`](https://github.com/ZhengyiLuo/PULSE/blob/b8e12b29f1e7e952be3779a0528ed7bf2abc55bc/scripts/data_process/convert_amass_data.py#L26)
\* The pretrained model by imitation learning is also trained by myself, not using the model you released. We additionally verified that the released imitation model(`pulse_vae_iclr/Humanoid.pth`) also cannot reproduce the reported result.

Seems the evaluation results on training set match with the paper. Is overfitting might be the possible problem?

\+ The recent commit uses `auto_pmcp_soft` turned on, but I can't locate where that flag is actually used to automatic hard-negative mining procedure.

\+ Could you also kindly share the pretrained weight for the VR tracking result? I want to see if my current evaluation code reproduces similar result for the test set as reported in the paper.

**Comment by ZhengyiLuo** [comment]
> Thank you for answering!
> 
> #### 1. How long is your model trained for?
> I trained with previous config (1536 envs), so total 40,000 epochs to match 2 x 10^9 samples.
> 
> #### 2. What is the training set success rate for your tracking model right now?
> success rate	g-mpjpe	mpjpe	acc	vel
> Train set	0.96213	59.30511	51.59011	4.02125	6.09905
> Validation set	0.8	84.7218	71.78029	6.72136	10.5008
> Evaluation set	0.67391	116.4351	81.12007	8.69799	13.62242
> * All 3 sets are created with [convert_amass_data.py](https://github.com/ZhengyiLuo/PULSE/blob/master/scripts/data_process/convert_amass_data.py), changing [`process_split`](https://github.com/ZhengyiLuo/PULSE/blob/b8e12b29f1e7e952be3779a0528ed7bf2abc55bc/scripts/data_process/convert_amass_data.py#L26) * The pretrained model by imitation learning is also trained by myself, not using the model you released. We additionally verified that the released imitation model(`pulse_vae_iclr/Humanoid.pth`) also cannot reproduce the reported result.
> 
> Seems the evaluation results on training set match with the paper. Is overfitting might be the possible problem?
> 
> + The recent commit uses `auto_pmcp_soft` turned on, but I can't locate where that flag is actually used to automatic hard-negative mining procedure.
> 
> + Could you also kindly share the pretrained weight for the VR tracking result? I want to see if my current evaluation code reproduces similar result for the test set as reported in the paper.

Sorry for the late reply! Had a busy month of job hunting. I just uploaded the `auto_pmcp_soft` corresponding code; this part is pretty important; the 96.2% seems to be on-part with one trained without it set. Could you share the training curve with wandb (the success rate). Turning the hard-mining on is pretty important. 

I have uploaded the model reported in the paper here: https://drive.google.com/file/d/1Emee_-ls7gBDrUHJulZOlIeLdk6jCEkH/view?usp=drive_link

you should be able to run the evaluation code like this
```
python phc/run_hydra.py env.task=HumanoidImZ env=env_pulse_im exp_name=pulse_vr robot.real_weight_porpotion_boxes=False learning=pulse_z_vr env.models=['output/HumanoidIm/pulse_vae_iclr/Humanoid.pth'] env.motion_file=data/amass/pkls/amass_isaac_im_test_take6_upright_slim.pkl epoch=25500 im_eval=True test=True headless=True env.num_envs=150
```

These are the results I ran this morning on the test set. IsaacGym is non-deterministic, so there are variations from run to run. 

------------------------------------------
Success Rate: 0.9270072993
All:  mpjpe_g: 94.451   mpjpe_l: 71.178         mpjpe_pa: 54.292        accel_dist: 8.809       vel_dist: 14.997
Succ:  mpjpe_g: 87.920  mpjpe_l: 67.427         mpjpe_pa: 51.399        accel_dist: 7.563       vel_dist: 13.649
output/HumanoidIm/pulse_vr
